lib/backend/k8s/conversion.go

```diff
@@ -0,0 +1,359 @@
+		return nil, err
+	}
+	return &model.KVPair{
+		Key:      model.PoolKey{CIDR: pool.CIDR},
+		Value:    &pool,
+		Revision: ns.ObjectMeta.ResourceVersion,
+	}, nil
+}
+
+func (c converter) namespaceToProfile(ns *k8sapi.Namespace) (*model.KVPair, error) {
+	// Determine the ingress action based off the DefaultDeny annotation.
+	ingressAction := "allow"
+	for k, v := range ns.ObjectMeta.Annotations {
+		if k == policyAnnotation {
+			np := namespacePolicy{}
+			if err := json.Unmarshal([]byte(v), &np); err != nil {
+				return nil, err
+			}
+			if np.Ingress.Isolation == "DefaultDeny" {
+				ingressAction = "deny"
+			}
+		}
+	}
+
+	// Generate the labels to apply to the profile.
+	labels := map[string]string{}
+	for k, v := range ns.ObjectMeta.Labels {
+		labels[fmt.Sprintf("k8s_ns/label/%s", k)] = v
+	}
+
+	name := fmt.Sprintf("k8s_ns.%s", ns.ObjectMeta.Name)
+	kvp := model.KVPair{
+		Key: model.ProfileKey{Name: name},
+		Value: &model.Profile{
+			Rules: model.ProfileRules{
+				InboundRules:  []model.Rule{model.Rule{Action: ingressAction}},
+				OutboundRules: []model.Rule{model.Rule{Action: "allow"}},
+			},
+			Tags:   []string{name},
+			Labels: labels,
+		},
+		Revision: ns.ObjectMeta.ResourceVersion,
+	}
+	return &kvp, nil
+}
+
+func (c converter) isHostNetworked(pod *k8sapi.Pod) bool {
+	return pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.HostNetwork == true
+}
+
+func (c converter) hasIPAddress(pod *k8sapi.Pod) bool {
+	return pod.Status.PodIP != ""
+}
+
+func (c converter) podToWorkloadEndpoint(pod *k8sapi.Pod) (*model.KVPair, error) {
+	// Pull out the profile and workload ID based on pod name and Namespace.
+	profile := fmt.Sprintf("k8s_ns.%s", pod.ObjectMeta.Namespace)
+	workload := fmt.Sprintf("%s.%s", pod.ObjectMeta.Namespace, pod.ObjectMeta.Name)
+
+	// If the pod doesn't have an IP address yet, then it hasn't gone through CNI.
+	// Treat this as if it didn't exist.
+	ipNets := []cnet.IPNet{}
+	if c.hasIPAddress(pod) {
+
+		// Parse the Pod's IP address.
+		_, ipNet, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", pod.Status.PodIP))
+		if err != nil {
+			return nil, err
+		}
+		ipNets = []cnet.IPNet{*ipNet}
+	}
+
+	// Generate the interface name and MAC based on workload.  This must match
+	// the host-side veth configured by the CNI plugin.
+	interfaceName, mac := VethParamsForWorkload(workload)
+
+	// Build the labels map.
+	labels := pod.ObjectMeta.Labels
+	labels["calico/k8s_ns"] = pod.ObjectMeta.Namespace
+
+	// TODO: Determine plan for `Hostname` - what value do we want here?
+	// It could be the Node IP addr, or it could be the NodeName.
+
+	// Create the key / value pair to return.
+	kvp := model.KVPair{
+		Key: model.WorkloadEndpointKey{
+			Hostname:       pod.Spec.NodeName,
+			OrchestratorID: "k8s",
+			WorkloadID:     workload,
+			EndpointID:     "eth0",
+		},
+		Value: &model.WorkloadEndpoint{
+			State:      "active",
+			Name:       interfaceName,
+			Mac:        &cnet.MAC{HardwareAddr: mac},
```

Comment:
Ooh, that's nice.

If it's omitted, then it just doesn't do the src MAC check in iptables?